### PR TITLE
Notifications

### PR DESCRIPTION
**Pull request by ozasadnyy**

### CODE REVIEW

File: scripts/redux/actions.js [hunk 1]
```diff
@@ -420,6 +425,105 @@ const subscribeActions = {
...
+      }
+    }
+  })
+});
+messaging.onTokenRefresh(() => {
+  notificationsActions.getToken(true);
+});
+const notificationsActions = {
+  requestPermission: () => {
+    return messaging.requestPermission()
```
Comment (sophieH29) [inline scripts/redux/actions.js]: doesn't requestPermision return token?

File: scripts/redux/actions.js [hunk 2]
```diff
@@ -420,6 +425,105 @@ const subscribeActions = {
...
+
+  getToken: subscribe => {
+    return messaging.getToken()
+      .then(currentToken => {
+        if (currentToken) {
+          const state = store.getState();
+          const subscribersRef = firebase.database().ref(`/notifications/subscribers/${currentToken}`);
+          const subscribersPromise = subscribersRef.once('value');
+
+          const userUid = state.user && state.user.uid ? state.user.uid : null;
```
Comment (sophieH29) [inline scripts/redux/actions.js]: state.user && (state.user.uid || null)

File: scripts/redux/actions.js [hunk 3]
```diff
@@ -420,6 +425,105 @@ const subscribeActions = {
...
+          let userSubscriptionsRef;
+          if (userUid) {
+            userSubscriptionsRef = firebase.database().ref(`/notifications/users/${userUid}/${currentToken}`);
+            userSubscriptionsPromise = userSubscriptionsRef.once('value');
+          }
+
+          Promise.all([subscribersPromise, userSubscriptionsPromise])
+            .then(results => {
+              const isSubscribed = results[0].val();
+              const isUserSubscribed = results[1] ? results[1].val() : false;
```
Comment (sophieH29) [inline scripts/redux/actions.js]: what is the difference between isSubscribed and isUserSubscribed?

File: src/hoverboard-app.html [hunk 4]
```diff
@@ -628,6 +674,31 @@ <h3 class="location">{$ location.short $}</h3>
       _removeQueryParams() {
         this.set('queryParams', null);
       }
+
+      _toggleNotifications() {
+        this._closeNotificationMenu();
+        if (this.notifications.status === NOTIFICATIONS_STATUS.GRANTED) {
+          notificationsActions.unsubscribe(this.notifications.token);
```
Comment (sophieH29) [inline src/hoverboard-app.html]: why you unsubscribe when permission is granted?

File: scripts/redux/actions.js (hunk 1 above)
Comment (ozasadnyy) [reply scripts/redux/actions.js]: no, https://firebase.google.com/docs/reference/js/firebase.messaging.Messaging#requestPermission

File: src/hoverboard-app.html (hunk 4 above)
Comment (ozasadnyy) [reply src/hoverboard-app.html]: if people get rid of notifications and prevent them blocking notifications on the website

File: scripts/redux/actions.js (hunk 3 above)
Comment (ozasadnyy) [reply scripts/redux/actions.js]: isSubscribed - for anonymous users
isUserSubscribed - authenticated

Comment (sophieH29) [reply scripts/redux/actions.js]: Then I suggest to rename them like
isAnonymousUserSubscribed
isAuthenticatedUserSubscribed

Or any other your variant, just to make code more obvious

File: functions/schedule-notifications.js [hunk 5]
```diff
@@ -51,19 +49,24 @@ const sendPushNotificationToUsers = (userIds, payload) => {
...
+  const schedulePromise = admin.database().ref(`/schedule`).once('value');
 
-  return admin.database().ref(`/schedule`).once('value')
-    .then(scheduleSnapshot => {
+  return Promise.all([notificationsConfigPromise, schedulePromise])
+    .then(([notificationsConfigSnapshot, scheduleSnapshot]) => {
+      const notificationsConfig = notificationsConfigSnapshot.val();
       const schedule = scheduleSnapshot.val();
+      console.log('notificationsConfig', notificationsConfig);
+      console.log('schedule', schedule);
```
Comment (sophieH29) [inline functions/schedule-notifications.js]: do we need to console log this info, or you forgot to remove after debugging?

File: functions/schedule-notifications.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
+'use strict';
+
+const functions = require('firebase-functions');
+const admin = require('firebase-admin');
+const moment = require('moment');
+
+const FORMAT = 'HH:mm';
+const BASE_URL = 'https://hoverboard-v2-dev.firebaseapp.com';
```
Comment (sophieH29) [inline functions/schedule-notifications.js]: better move it to config also

Comment (ozasadnyy) [reply functions/schedule-notifications.js]: It's not finished, working on another solution

File: functions/schedule-notifications.js (hunk 5 above)
Comment (ozasadnyy) [reply functions/schedule-notifications.js]: no,  forgot to remove ;)

File: functions/schedule-notifications.js (hunk 6 above)
Comment (sophieH29) [reply functions/schedule-notifications.js]: ok

File: database.rules.json [hunk 7]
```diff
@@ -18,14 +18,6 @@
         ".write": "auth.uid === $uid"
       }
     },
-    "featuredSessions": {
```
Comment (sophieH29) [inline database.rules.json]: why have you removed the rules?

Comment (ozasadnyy) [reply database.rules.json]: Because it was the duplication

**Pull request merged**